Lib/pathlib.py

```diff
@@ -320,8 +320,9 @@ def _from_parsed_parts(cls, drv, root, parts):
     def _format_parsed_parts(cls, drv, root, parts):
         if drv or root:
             return drv + root + cls._flavour.sep.join(parts[1:])
-        else:
-            return cls._flavour.sep.join(parts)
+        elif parts and cls._flavour.splitdrive(parts[0])[0]:
```

Comment:
```suggestion
        elif parts and cls._flavour.splitroot(parts[0])[0]:
```

Reply:
Slower on Linux, as `posixpath.splitdrive()` is trivial whereas `posixpath.splitroot()` is not!

Reply:
We should probably implement a `_Py_splitroot(wchar_t *path, wchar_t **root, wchar_t **rest)` function in "Python/fileutils.c", to be used by `[nt|posix]._path_splitroot()` and used internally by `_Py_normpath()`. Then `cls._flavour.splitroot()` would be relative cheap on both POSIX and Windows.

There are a couple of `nt._path_splitroot()` calls in `importlib`. They would have to be updated to support a three-part split `(drive, root, rest)` instead of the two-part split `(anchor, rest)`.

Reply:
Maybe it's worth adding a `hasroot` method that the different flavours can implement however is most efficient? We don't actually _need_ the root here, just to determine whether there is one.

Or maybe some kind of `does_this_join_as_a_filename` or `can_join_without_removing_previous_segments` API?

Reply:
Just in case it isn't obvious why a root can't appear: the `_parts` list is a list of `os.path.sep`-delimited path parts after the drive and root. And as path roots can only consist of `os.path.sep` (and `altsep` on Windows), it follows that no element in `_parts` can contain a root.

Reply:
> We only care about the drive, and so splitdrive(path)[0] works well enough.

I suggested calling `splitroot()` directly only because it's the basis for `ntpath.splitdrive()`. I didn't consider the tradeoff that `splitdrive()` is cheap on POSIX. If `splitroot()` were cheap on POSIX, I assume you would be okay with using it here. This can wait until `splitroot()` is implemented in C.

I think `splitroot()` warrants a C implementation since it's a required step in our basic path handling on Windows -- `join()`, `split()`, `relpath()`, and `commonpath()`. Speeding it up gives a little boost across the board. Also, it would be less confusing if `nt._path_splitroot()` actually implemented `ntpath.splitroot()`. 

If implemented, using `_Py_splitroot()` in `_Py_normpath()` would help to ensure consistency. Currently, for example, `ntpath.normpath('//?/UNC/server/share/../..')` is correct on POSIX but wrong on Windows because `_Py_normpath()` incorrectly handles "//?/UNC/" as the root instead of "//?/UNC/server/share/".

Reply:
I've logged an issue to tackle a C implementation of `splitroot()`: 

- #102511